turbo/engineapi/request_list.go

```diff
@@ -143,34 +151,48 @@ func (rl *RequestList) WaitForRequest(onlyNew bool, noWait bool) (interrupt Inte
 }
 
 func (rl *RequestList) IsWaiting() bool {
-	return atomic.LoadUint32(&rl.waiting) != 0
+	return rl.waiting.Load() != 0
 }
 
 func (rl *RequestList) updateWaiting(val uint32) {
-	rl.waiterMtx.Lock()
-	defer rl.waiterMtx.Unlock()
-	atomic.StoreUint32(&rl.waiting, val)
+	rl.waiting.Store(val)
 
 	if val == 1 {
-		// something might be waiting to be notified of the waiting state being ready
-		for _, c := range rl.waiters {
-			c <- struct{}{}
-		}
-		rl.waiters = make([]chan struct{}, 0)
+		rl.notifyWaiters()
 	}
 }
 
-func (rl *RequestList) WaitForWaiting(c chan struct{}) {
+func (rl *RequestList) WaitForWaiting(ctx context.Context) {
 	rl.waiterMtx.Lock()
-	defer rl.waiterMtx.Unlock()
-	val := atomic.LoadUint32(&rl.waiting)
-	if val == 1 {
-		// we are already waiting so just send to the channel and quit
-		c <- struct{}{}
+	isWaiting := rl.waiting.Load()
+	if isWaiting == 1 {
+		// we are already waiting so just return
+		rl.waiterMtx.Unlock()
```

Comment:
I advise to unlock mutex by `defer`, otherwise `panic()` may cause deadlock

Reply:
also: why need mutex lock here? `rl.waiting` is atomic already

Reply:
This has been the problem I'm facing, if I check the state of waiting it can be changed immediately afterwards whilst the waiter is created and it will never be notified.  So we end up waiting the full timeout time when we could have waited just 5ms or something.  The lock here is to ensure that whilst the state of `rl.waiting` is being checked that it can't be updated until we've returned from this function or that the waiter is in place to monitor for the update